File: code/python/proxy/ProxyLegalService.py
from LegalService import LegalService
from RealLegalService import RealLegalService

class ProxyLegalService(LegalService):
    def __init__(self):
        self.real_legal_service = None

    def request_legal_assistance(self):
        if self.real_legal_service is None:
            self.real_legal_service = RealLegalService()

        print("Proxy legal service is requesting legal assistance.")
        self.real_legal_service.provide_legal_assistance()

File: code/typescript/src/proxy/ProxyLegalService.ts
import { LegalService } from "./LegalService";
import { RealLegalService } from "./RealLegalService";

export class ProxyLegalService implements LegalService {
    private realLegalService?: RealLegalService;

    requestLegalAssistance(): void {
        if (!this.realLegalService) {
            this.realLegalService = new RealLegalService();
        }

        console.log("Proxy legal service is requesting legal assistance.");
        this.realLegalService.provideLegalAssistance();
    }
}
